controllers/grafananotificationtemplate_controller.go

```diff
@@ -0,0 +1,214 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/go-logr/logr"
+	"github.com/grafana/grafana-openapi-client-go/client/provisioning"
+	"github.com/grafana/grafana-openapi-client-go/models"
+	grafanav1beta1 "github.com/grafana/grafana-operator/v5/api/v1beta1"
+	client2 "github.com/grafana/grafana-operator/v5/controllers/client"
+)
+
+const (
+	conditionNotificationTemplateSynchronized = "NotificationTemplateSynchronized"
+)
+
+// GrafanaNotificationTemplateReconciler reconciles a GrafanaNotificationTemplate object
+type GrafanaNotificationTemplateReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationtemplates,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationtemplates/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationtemplates/finalizers,verbs=update
+
+func (r *GrafanaNotificationTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	controllerLog := log.FromContext(ctx).WithName("GrafanaNotificationTemplateReconciler")
+	r.Log = log.FromContext(ctx)
+
+	notificationTemplate := &grafanav1beta1.GrafanaNotificationTemplate{}
+	err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}, notificationTemplate)
+	if err != nil {
+		if kuberr.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		controllerLog.Error(err, "Failed to get GrafanaNotificationTemplate")
+		return ctrl.Result{RequeueAfter: RequeueDelay}, err
+	}
+
+	if notificationTemplate.GetDeletionTimestamp() != nil {
+		if controllerutil.ContainsFinalizer(notificationTemplate, grafanaFinalizer) {
+			err := r.finalize(ctx, notificationTemplate)
+			if err != nil {
+				return ctrl.Result{RequeueAfter: RequeueDelay}, fmt.Errorf("failed to finalize GrafanaNotificationTemplate: %w", err)
+			}
+			controllerutil.RemoveFinalizer(notificationTemplate, grafanaFinalizer)
+			if err := r.Update(ctx, notificationTemplate); err != nil {
+				r.Log.Error(err, "failed to remove finalizer")
+				return ctrl.Result{}, fmt.Errorf("failed to update GrafanaNotificationTemplate: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	defer func() {
+		notificationTemplate.Status.LastResync = metav1.Time{Time: time.Now()}
+		if err := r.Client.Status().Update(ctx, notificationTemplate); err != nil {
+			r.Log.Error(err, "updating status")
+		}
+		if meta.IsStatusConditionTrue(notificationTemplate.Status.Conditions, conditionNoMatchingInstance) {
+			controllerutil.RemoveFinalizer(notificationTemplate, grafanaFinalizer)
+		} else {
+			controllerutil.AddFinalizer(notificationTemplate, grafanaFinalizer)
+		}
+		if err := r.Update(ctx, notificationTemplate); err != nil {
+			r.Log.Error(err, "failed to set finalizer")
+		}
```

Comment:
Finalizers need to be updated here especially as this fixes the config drift:
```suggestion
		if meta.IsStatusConditionTrue(notificationTemplate.Status.Conditions, conditionNoMatchingInstance) {
			if err := removeFinalizer(ctx, r.Client, notificationTemplate); err != nil {
				r.Log.Error(err, "failed to remove finalizer")
			}
		} else {
			if err := addFinalizer(ctx, r.Client, notificationTemplate); err != nil {
				r.Log.Error(err, "failed to set finalizer")
			}
		}
```